test/validation/api/ipsec/ipsec_inline_reass.c

```diff
@@ -0,0 +1,56 @@
+ /* Copyright (c) 2020, Marvell
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:	 BSD-3-Clause
+ */
+
+#include "ipsec.h"
+
+static int ipsec_inline_reass_sync_init(odp_instance_t *inst)
+{
+	int rc;
+
+	suite_context.inbound_op_mode = ODP_IPSEC_OP_MODE_INLINE;
+	suite_context.outbound_op_mode = ODP_IPSEC_OP_MODE_SYNC;
+	suite_context.inline_in_reass_ipv4 = true;
+
+	rc = ipsec_init(inst, ODP_IPSEC_OP_MODE_INLINE);
+	if (rc != 0)
+		return rc;
+
+	suite_context.pool = odp_pool_lookup("packet_pool");
+	if (suite_context.pool == ODP_POOL_INVALID)
+		return -1;
+	suite_context.queue = odp_queue_lookup("ipsec-out");
+	if (suite_context.queue == ODP_QUEUE_INVALID)
+		return -1;
+	suite_context.pktio = odp_pktio_lookup("loop");
+	if (suite_context.pktio == ODP_PKTIO_INVALID)
+		return -1;
+
+	return ipsec_config(*inst);
+}
+
+odp_suiteinfo_t reass_ipsec_suites[] = {
+	{"IPsec-inline-reass", ipsec_suite_init, ipsec_in_term,
+	 ipsec_in_reass_suite},
+	ODP_SUITE_INFO_NULL,
+};
```

Comment:
I do not see why reassembly tests need a new suite and a test app, as opposed to just having additional test cases in the existing suites. Since the new suite used inline mode for inbound, reassembly would not get tested for async mode.

Reply:
@JannePeltonen Actually we could add this in the existing tests, but then we will not be able to test without reassembly. IOW, once we add reassembly to these UTs, the configure step will have reassembly enabled, and the non-reassembly path won't be tested.

Just to share some of our future plans, we intend to add following UTs
1. inline_out_tm (Once TM spec is merged)
2. inline_in_vector (in pipeline)

In all the above cases, it is better to test for all combinations. Do you have any suggestions on how we proceed?

I've anyway added the async, inline capability which I had described on the call. Irrespective of what we decide on UT, that would be required.

Reply:
@JannePeltonen I might have a solution. Something like,

```
+static int ipsec_in()
+{
+       int ret;
+
+       ret = ipsec_config(glbl_inst);
+       if (ret)
+               return ret;
+
+       return ipsec_suite_init();
+}
+
+static int ipsec_in_reass()
+{
+       int ret;
+
+       suite_context.inline_in_reass_ipv4 = true;
+
+       ret = ipsec_config(glbl_inst);
+       if (ret)
+               return ret;
+
+       return ipsec_suite_init();
 }

 odp_suiteinfo_t ipsec_suites[] = {
-       {"IPsec-in", ipsec_suite_init, ipsec_in_term, ipsec_in_suite},
+       {"IPsec-in", ipsec_in, ipsec_in_term, ipsec_in_suite},
+       {"IPsec-in-reass", ipsec_in_reass, ipsec_in_term, ipsec_in_suite},
        ODP_SUITE_INFO_NULL,
 };
```
But there is a major problem. Spec says odp_ipsec_config() should not be called more than once.
https://opendataplane.github.io/odp/group__odp__ipsec.html#ga85cc7deadb7406d340b6597fe4b516cc

Should we amend the above?

Reply:
> Actually we could add this in the existing tests, but then we will not be able to test without reassembly. IOW, once we add reassembly to these UTs, the configure step will have reassembly enabled, and the non-reassembly path won't be tested.

I do not see it that way. We can enable reassembly (basically hint that we might be using it) in the global IPsec config but still select on per-SA basis whether it is truly enabled. So we can test both with and without reassembly. I do not think there is necessarily a need to have global configs with both reassembly enabled and disabled in the API validation tests.

Reply:
Understood. I'll add reassembly tests to in-suite then. Thanks